### ISSUE TITLE
Contribsys repo intermediary cert expired - bundle install fails with Could not verify the SSL certificate for https://gems.contribsys.com/

### ISSUE THREAD

**Issue by tcdachille** [mperham/sidekiq#5008]
Similar to https://github.com/mperham/sidekiq/issues/4583, bundle install is failing with 

```
Retrying fetcher due to error (2/6): Bundler::Fetcher::CertificateFailureError Could not verify the SSL certificate for https://gems.contribsys.com/.    
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.    /usr/lib64/ruby/2.6.0/net/protocol.rb:44: warning: exception in verify_callback is ignored
```

Trying https://www.ssllabs.com/ssltest/analyze.html?d=gems.contribsys.com  reveals an expired cert 
![image](https://user-images.githubusercontent.com/14118064/135488373-44eb4882-0c2f-4390-a524-1eb8754776a4.png)

**Comment by mperham** [comment]
The certificate seems to work with `curl`, Net::HTTP and Safari without issue.

```ruby
> require 'net/https'
> Net::HTTP.get(URI("https://gems.contribsys.com"))
=> "This is Contributed Systems' Sidekiq Pro repo.\n"

**Comment by mperham** [comment]
https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/

**Comment by bpt2410** [comment]
We got the same issue

**Comment by mperham** [comment]
tl;dr: update ruby and openssl.

**Comment by mperham** [comment]
https://community.letsencrypt.org/t/openssl-client-compatibility-changes-for-let-s-encrypt-certificates/143816

> The behavior that causes this incompatibility was fixed over 4 years ago with the release of OpenSSL 1.1.0.

**Comment by alejandrovelez7** [comment]
Updating bundler to >= 2.2 may be useful as well since older versions were using different ssl handling https://bundler.io/blog/2020/12/09/bundler-v2-2.html

**Comment by LizhangX** [comment]
Same here, using jruby docker image jruby:9.2.14.0-jdk.
Openssl is `OpenSSL 1.1.0f  25 May 2017`
But still getting the same issue during bundle install.

**Comment by schmiddy** [comment]
FYI we see this error on the latest Amazon Linux 2 AMI, you can check by launching an EC2 instance from public AMI `ami-0c0c5692729d9c263` in us-west-2 (AMI Name `amzn2-ami-kernel-5.10-hvm-2.0.20210813.1-x86_64-gp2`). 

```
[ec2-user@ip-172-31-40-244 ~]$ openssl version
OpenSSL 1.1.1  11 Sep 2018 (Library: OpenSSL 1.1.1g FIPS  21 Apr 2020)

[ec2-user@ip-172-31-40-244 ~]$ bundle --version
Bundler version 2.2.28

[ec2-user@ip-172-31-40-244 ~]$ ruby --version
ruby 2.6.8p205 (2021-07-07 revision 67951) [x86_64-linux]
```

**Comment by mperham** [comment]
https://community.letsencrypt.org/t/help-thread-for-dst-root-ca-x3-expiration-september-2021/149190/214

**Comment by davebenson** [comment]
It appears to be a problem on google appengine's flex ruby environment as well.  guess i'm SOL until google fixes it, unless i want to switch to a custom environment and build my own docker image.  ugh.

**Comment by dannyfallon** [comment]
On Amazon Linux 2 we worked around this today by punting the expired certificate into the PKI blacklist on our machines:

```
trust dump --filter "pkcs11:id=%C4%A7%B1%A4%7B%2C%71%FA%DB%E1%4B%90%75%FF%C4%15%60%85%89%10;type=cert" > /etc/pki/ca-trust/source/blacklist/dst-root.p11-kit && update-ca-trust extract
```

**Comment by sm-kwohlhueter** [comment]
We're using an AmazonLinux2 container for AWS CodeBuild, and @dannyfallon's one line'r fixed it for us.
🥇 for @dannyfallon

**Comment by vishalzambre** [comment]
> We're using an AmazonLinux2 container for AWS CodeBuild, and @dannyfallon's one line'r fixed it for us.
> 🥇 for @dannyfallon

which one liner? I'm also using AmazonLinux2 with EBS, can you please help

**Comment by sm-kwohlhueter** [comment]
`trust dump --filter "pkcs11:id=%C4%A7%B1%A4%7B%2C%71%FA%DB%E1%4B%90%75%FF%C4%15%60%85%89%10;type=cert" > /etc/pki/ca-trust/source/blacklist/dst-root.p11-kit && update-ca-trust extract`

**Comment by vishalzambre** [comment]
It is not working on

```
NAME="Amazon Linux AMI"
VERSION="2018.03"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2018.03"
PRETTY_NAME="Amazon Linux AMI 2018.03"
ANSI_COLOR="0;33"
CPE_NAME="cpe:/o:amazon:linux:2018.03:ga"
HOME_URL="http://aws.amazon.com/amazon-linux-ami/"
```

**Comment by vishalzambre** [comment]
Any other alternative?

**Comment by sm-kwohlhueter** [comment]
Try using AmazonLinux2. You seem to be on AmazonLinux.

**Comment by vishalzambre** [comment]
Any other workaround for AmazonLinux?

**Comment by dmt219** [comment]
@vishalzambre This works for me:
1. Run `trust dump --filter "pkcs11:id=%C4%A7%B1%A4%7B%2C%71%FA%DB%E1%4B%90%75%FF%C4%15%60%85%89%10;type=cert"` in another Linux environment which supports the `trust` tool (Amazon Linux 1 doesn't have this)
2. Paste the result into
`/etc/pki/ca-trust/source/blacklist/dst-root.p11-kit`
3. Run `update-ca-trust extract`

**Comment by vishalzambre** [comment]
I'm getting now `p11-kit: dst-root.p11-kit: nss-mozilla-ca-policy: invalid or unsupported attribute`

**Comment by vishalzambre** [comment]
Now I can see 

```
# update-ca-trust extract
p11-kit: dst-root.p11-kit: pkcs11: attribute before p11-kit section header
p11-kit: dst-root.p11-kit: pkcs11: attribute before p11-kit section header
p11-kit: dst-root.p11-kit: pkcs11: attribute before p11-kit section header
p11-kit: dst-root.p11-kit: pkcs11: attribute before p11-kit section header
p11-kit: dst-root.p11-kit: pkcs11: attribute before p11-kit section header
```

But still bundle install failing

```
Could not verify the SSL certificate for https://gems.contribsys.com/.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see
http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.

```

**Comment by sm-kwohlhueter** [comment]
Is the ISRG Root X1 CA in your cert store?

On AmazonLinux2 it's at `/etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt`. This could be different on AmazonLinux1.
When I run `/etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt  | grep -i "ISRG Root X1"` I get 1 result back.

**Comment by vishalzambre** [comment]
I can see

```
cat /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt  | grep -i "ISRG Root X1"
# ISRG Root X1
```

**Comment by sm-kwohlhueter** [comment]
That's good. 
Looking at this thread there are 3 options:
- find the AmazonLinux commands to blacklist the expired cert. (similar solution works on AmazonLinux2)
- update to a newer version of OpenSSL. `openssl version` should be at least `1.1.0`.
- with the ISRG Root X1 CA Cert in your cert store, hopefully when certbot renews the `*.contribsys.com` certificate in a few days the expired CA cert will fall off the cert chain and this issue will go away.

**Comment by dannyfallon** [comment]
Not sure you have any interest in doing this Mike, but from reading this evening I think that you can support older OpenSSL versions (and things based on those), at the expense of Android >4, <7.1.1 support if you generate a new certificate using the `--preferred-chain "ISRG Root X1"` argument for Certbot. There's probably less Android users of Contribsys servers than folks using the likes of Amazon Linux or old OpenSSL versions 😅

**Comment by bpt2410** [comment]
You can see this
https://aws.amazon.com/premiumsupport/knowledge-center/ec2-expired-certificate/

**Comment by vishalzambre** [comment]
Finally below snippet worked for me

```
# cp -i /etc/pki/tls/certs/ca-bundle.crt ~/ca-bundle.crt-backup

# trust dump --filter "pkcs11:id=%c4%a7%b1%a4%7b%2c%71%fa%db%e1%4b%90%75%ff%c4%15%60%85%89%10" | openssl x509 | sudo tee /etc/pki/ca-trust/source/blacklist/DST-Root-CA-X3.pem

# sudo update-ca-trust extract
```

**Comment by vishalzambre** [comment]
> You can see this
> https://aws.amazon.com/premiumsupport/knowledge-center/ec2-expired-certificate/

Ahh it is really nice

**Comment by adolfopeccin** [comment]
Any solution for applications hosted on AWS Elastic Beanstalk?

**Comment by bpt2410** [comment]
> Any solution for applications hosted on AWS Elastic Beanstalk?

You can use this command on AWS EB
`command: 'sudo sed -i"" "/Alias: DST Root CA X3/,/No Rejected Uses./d" /usr/share/pki/ca-trust-source/ca-bundle.trust.crt && sudo update-ca-trust'`

**Comment by tomchapin** [comment]
If you are using Rails on AWS Elastic Beanstalk, you have to implement the ssl fix before the gems are installed.

I created a file named "/.ebextensions/000_ssl_fix.config" with these contents:

```
commands:
  create_pre_dir:
    command: "mkdir -p /opt/elasticbeanstalk/hooks/preinit"
    ignoreErrors: true

files:
  "/opt/elasticbeanstalk/hooks/appdeploy/pre/00_fix_ssl.sh":
    mode: "000755"
    owner: root
    group: root
    content: |
      #!/usr/bin/env bash

      sudo sed -i"" "/Alias: DST Root CA X3/,/No Rejected Uses./d" /usr/share/pki/ca-trust-source/ca-bundle.trust.crt && sudo update-ca-trust
```

**Comment by DaveCollinsJr** [comment]
Having this same problem in Travis CI using "dist: xenial"  we are working through some solutions to this and will post if we find something.

**Comment by theist** [comment]
I think I fixed it in an old xenial CI docker container by: 
* Updating to the last `ca-certificates` 
* Editing `/etc/ca-certificates.conf`
* Placing a `!` before line `mozilla/DST_Root_CA_X3.crt`
* Running `update-ca-certificates`

After that I get this:
```
root@2d9486cf52f4:/app# ruby -v
ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]

root@2d9486cf52f4:/app# ruby
require 'net/https'
foo = Net::HTTP.get(URI("https://gems.contribsys.com"))
puts foo 
This is Contributed Systems' Sidekiq Pro repo.

root@2d9486cf52f4:/app# cat /etc/issue
Ubuntu 16.04.6 LTS \n \l

root@2d9486cf52f4:/app# openssl version
OpenSSL 1.0.2g  1 Mar 2016
```

But I haven't tried it yet for the real deal and there are a lot of moving parts when it comes to legacy systems, YMMV

**Comment by theist** [comment]
Reviewing my own preliminary solution I can confirm this issue is solved in my system (several legacy MRI rubies in ubuntu 16.04) by just adding these steeps as root:

```shell
sed 's/mozilla.DST_Root_CA_X3.crt/!mozilla\/DST_Root_CA_X3.crt/' -i /etc/ca-certificates.conf
update-ca-certificates 
```

Also worked in different combinations of debian / ubuntu systems and several MRI rubies.

**Comment by richkettle** [comment]
> Having this same problem in Travis CI using "dist: xenial" we are working through some solutions to this and will post if we find something.

@DaveCollinsJr We are having the same issues on travis. I've been in contact with them and nothing they have suggested works. We previously used `dist: trusty`. But changing it and adding various lines of `before_install` haven't worked yet. 

If anyone finds this looking for the travis issue what they suggested was:

```
dist: focal
before_install:
  - sudo apt-get update && sudo apt-get install apt-transport-https ca-certificates -y && sudo update-ca-certificates
```


or 

```
dist: xenial
before_install:
  - sudo apt update && sudo apt install libgnutls-openssl27 libgnutls30
```

The above haven't worked for me as of yet. I'm currently stripping down our `travis.yml` to recreate with the minimum.

**Comment by jordan-thoms** [comment]
@dandynaufaldi We're also seeing this issue on JRuby, and the ca-certificates.conf workaround doesn't help there as it's using its own openssl implementation

**Comment by dbackeus** [comment]
For the record: For those with issues on AWS Linux 2 I believe `sudo yum update` is enough to fix this issue at this point. Some updates to `openssl` and `ca-certificates` packages were released in the last week or so. At least I didn't personally have to run any other commands to fix SSL errors in our CI deployment.

**Comment by mperham** [comment]
I've removed the old, expired root reference. jruby-openssl should work now.

**Comment by KLForsythe** [comment]
@mperham You mentioned that the fix implemented  in https://github.com/mperham/sidekiq/issues/5008#issuecomment-944612480 should last until early December.  We are using sidekiq-pro, and our builds are newly failing due to this issue `Could not verify the SSL certificate for https://gems.contribsys.com/`   Can you implement the fix for your renewed certs?

**Comment by mperham** [comment]
@KLForsythe The cert has not changed since Oct 15th and no one else is reporting a problem. I'm giving one quarter for people to roll out any necessary client upgrades (LE certs renew every quarter). The current cert expires on Jan 5 2022 so it will be renewed very soon.

**Comment by dylanbromby** [comment]
Our deploys were working fine this morning, but some time after 10a PST, we started seeing `Could not verify the SSL certificate for https://gems.contribsys.com` errors during our deploys. We're actively investigating but posting here in case anyone else is experiencing and/or has a solution.

**Comment by j3ck** [comment]
mac os 12.1
```
sed 's/mozilla.DST_Root_CA_X3.crt/!mozilla\/DST_Root_CA_X3.crt/' -i /etc/ssl/cert.pem
SSL_CERT_FILE=/etc/ssl/cert.pem bundle install
```

**Comment by krzkrzkrz** [comment]
I just paid for a Pro version today. Unfortunately, I am seeing these errors when running `bundle install`:

```
$ bundle install
Fetching source index from http://gems.contribsys.com/

Retrying fetcher due to error (2/4): Bundler::Fetcher::CertificateFailureError Could not verify the SSL certificate for http://gems.contribsys.com/.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.
Retrying fetcher due to error (3/4): Bundler::Fetcher::CertificateFailureError Could not verify the SSL certificate for http://gems.contribsys.com/.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.
Retrying fetcher due to error (4/4): Bundler::Fetcher::CertificateFailureError Could not verify the SSL certificate for http://gems.contribsys.com/.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.
Could not verify the SSL certificate for http://gems.contribsys.com/.
There is a chance you are experiencing a man-in-the-middle attack, but most likely your system doesn't have the CA certificates needed for verification. For information about OpenSSL certificates, see
http://bit.ly/ruby-ssl. To connect without using SSL, edit your Gemfile sources and change 'https' to 'http'.
```

I tried changing `source "https://gems.contribsys.com/"` to `source "http://gems.contribsys.com/"`. Same error.

This is still an issue. Currently on Mac OSX Big Sur 11.2.3
I have `openssl` already installed via Homebrew (i.e. `brew install openssl`)

`Gemfile` has:

```
source "http://gems.contribsys.com/" do
  gem 'sidekiq-pro'
end
```

Any ideas on a fix?
